sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressionsJoni.scala

```diff
@@ -0,0 +1,481 @@
+          Since Spark 2.0, string literals are unescaped in our SQL parser. For example, in order
+          to match "\abc", the pattern should be "\\abc".
+          When SQL config 'spark.sql.parser.escapedStringLiterals' is enabled, it fallbacks
+          to Spark 1.6 behavior regarding string literal parsing. For example, if the config is
+          enabled, the pattern to match "\abc" should be "\abc".
+      * escape - an character added since Spark 3.0. The default escape character is the '\'.
+          If an escape character precedes a special symbol or another escape character, the
+          following character is matched literally. It is invalid to escape any other character.
+  """,
+  examples = """
+    Examples:
+      > SELECT _FUNC_('Spark', '_park');
+      true
+      > SET spark.sql.parser.escapedStringLiterals=true;
+      spark.sql.parser.escapedStringLiterals	true
+      > SELECT '%SystemDrive%\Users\John' _FUNC_ '\%SystemDrive\%\\Users%';
+      true
+      > SET spark.sql.parser.escapedStringLiterals=false;
+      spark.sql.parser.escapedStringLiterals	false
+      > SELECT '%SystemDrive%\\Users\\John' _FUNC_ '\%SystemDrive\%\\\\Users%';
+      true
+      > SELECT '%SystemDrive%/Users/John' _FUNC_ '/%SystemDrive/%//Users%' ESCAPE '/';
+      true
+  """,
+  note = """
+    Use RLIKE to match with standard regular expressions.
+  """,
+  since = "1.0.0")
+// scalastyle:on line.contains.tab
+case class LikeJoni(left: Expression, right: Expression, escapeChar: Char)
+  extends StringRegexExpressionJoni {
+
+  def this(left: Expression, right: Expression) = this(left, right, '\\')
+
+  override def escape(v: Array[Byte]): Array[Byte] = StringUtils.escapeLikeJoniRegex(v, escapeChar)
+
+  override def matches(regex: Regex, input: Array[Byte]): Boolean = {
+    regex.matcher(input).`match`(0, input.length, Option.DEFAULT) == input.size
+  }
+
+  final override val nodePatterns: Seq[TreePattern] = Seq(LIKE_FAMLIY)
+
+  override def toString: String = escapeChar match {
+    case '\\' => s"$left LIKEJoni $right"
+    case c => s"$left LIKEJoni $right ESCAPE '$c'"
+  }
+
+  override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val regexClass = classOf[Regex].getName
+    val optionClass = classOf[Option].getName
+    val encodingClass = classOf[UTF8Encoding].getName
+    val syntaxClass = classOf[Syntax].getName
+    val escapeFunc = StringUtils.getClass.getName.stripSuffix("$") + ".escapeLikeJoniRegex"
+    val regex = ctx.freshName("regex")
+    val patternName = ctx.freshName("pattern")
+
+    if (right.foldable) {
+      val rVal = right.eval()
+      if (rVal != null) {
+        val tmp =
+          StringEscapeUtils.escapeJava(
+            new String(escape(rVal.asInstanceOf[UTF8String].getBytes), "utf-8"))
+        val pattern = ctx.addMutableState(regexClass, regex,
+          v => s"""
+                      byte[] $patternName = UTF8String.fromString("${tmp}").getBytes();
```

Comment:
```sql
select 
  case when columnA rlike '.*abcd.*' then 1 
       when columnA rlike '.*abcdc.*' then 2  else 0 end as a
from tableA
```
codegen exception:
org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 32, Column 8: Redefinition of local variable "pattern"

Reply:
hmm... which Spark version are you using? I cannot reproduce your issue in the master branch @SparksFyz

Reply:
3.2.0

Reply:
Spark 3.2.0 use janino 3.0.16, this is different from the master, it looks more like a issue of janino

Reply:
@SparksFyz Could you send the codegen code?